### PR TITLE
Ability to lift a ScalaVersion from a String

### PR DESCRIPTION
**Pull request by joan38**
Hi,

I'm helping @DavidGregory084 to maintain the [mill-tpolecat](https://github.com/DavidGregory084/mill-tpolecat) and I just found out about this project and the split with [sbt-tpolecat](https://github.com/typelevel/sbt-tpolecat). Great idea! and thank you for doing this!

Now I feel like we could deprecate mill-tpolecat and only use this dependency in mill. Here is an example:
```
import $ivy.`org.typelevel::scalac-options:0.1.4`

object `my-project` extends ScalaModule {
  override def scalaVersion = "3.3.1"
  override def scalacOptions = ScalacOptions.defaultTokensForVersion(ScalaVersion.V3_3_0)
}
```
You can see this is very compact already and mill-tpolecat would not really add much value.

However the thing that bothers me is that now we have to edit the Scala version at multiple places.
With this PR it would look like this:
```
import $ivy.`org.typelevel::scalac-options:0.1.4`

object `my-project` extends ScalaModule {
  override def scalaVersion = "3.3.1"
  override def scalacOptions = ScalacOptions.defaultTokensForVersion(ScalaVersion.unsafeFromString(scalaVersion()))
}
```

Thanks

### CODE REVIEW

File: lib/src/main/scala/org/typelevel/scalacoptions/ScalaVersion.scala [hunk 1]
```diff
@@ -44,6 +44,14 @@ object ScalaVersion {
   val V3_3_0   = ScalaVersion(3, 3, 0)
   val V3_3_1   = ScalaVersion(3, 3, 1)
 
+  def apply(version: String): ScalaVersion = {
+    val regex = raw"""(\d+)\.(\d+)\.(\d+).*""".r
+    version match {
+      case regex(major, minor, patch) => ScalaVersion(major.toLong, minor.toLong, patch.toLong)
+      case _                          => throw new IllegalArgumentException(s"Scala version $version not recognized")
+    }
+  }
```
[the remark this answers is not in the document]
Comment (joan38) [reply lib/src/main/scala/org/typelevel/scalacoptions/ScalaVersion.scala:53]: Thanks for the comment.
Sounds like @armanbilge is agreeing?

So would that mean we could not upgrade to a newer Scala version until this dependency is supporting it? Might me ok to me. It's kind of the same with semanticdb. WDYT?

[the remark this answers is not in the document]
Comment (armanbilge) [reply lib/src/main/scala/org/typelevel/scalacoptions/ScalaVersion.scala:53]: Sorry, let me clarify: I am 👍 to `fromString` and `Option` (or `Either`) but I think validation against a list of known versions is not worth the maintenance effort i.e. it's not buying us much and its making a lot of busywork to re-publish for every Scala release.

[the remark this answers is not in the document]
Comment (satorg) [reply lib/src/main/scala/org/typelevel/scalacoptions/ScalaVersion.scala:53]: Yes, it was on the question side. Feel free to disregard the validation please. Perhaps, I overthought it.

[the remark this answers is not in the document]
Comment (joan38) [reply lib/src/main/scala/org/typelevel/scalacoptions/ScalaVersion.scala]: I pushed an new version. Have a look.

File: lib/src/main/scala/org/typelevel/scalacoptions/ScalaVersion.scala [hunk 2]
```diff
@@ -44,6 +44,18 @@ object ScalaVersion {
   val V3_3_0   = ScalaVersion(3, 3, 0)
   val V3_3_1   = ScalaVersion(3, 3, 1)
 
+  def fromString(version: String): Either[IllegalArgumentException, ScalaVersion] = {
+    val regex = raw"""(\d+)\.(\d+)\.(\d+).*""".r
```
Comment (satorg) [inline lib/src/main/scala/org/typelevel/scalacoptions/ScalaVersion.scala:48]: I wonder, what is the last part `.*` in the regex is supposed for?
To me it looks like it's going to accept any arbitrary string after the last digit, like "1.2.3abc" – is that correct?

Another thought – perhaps it makes sense to keep the pre-compiled regex value in a private value outside of the method. To avoid getting it compiled every time upon the method call. Wdyt?

**Comment by joan38** [comment]
Are we good with this?

**Comment by joan38** [comment]
Do we need someone to merge the PR?

**Comment by joan38** [comment]
@satorg How can we move forward on this?

File: lib/src/main/scala/org/typelevel/scalacoptions/ScalaVersion.scala (hunk 2 above)
Comment (armanbilge) [reply lib/src/main/scala/org/typelevel/scalacoptions/ScalaVersion.scala]: > but you are right here it's a bug and would blow on the `patch.toLong`

Wait, why is that? The three parenthetical groups in the regex are the major, minor, and patch, the other stuff would be omitted, no?

I hate to say it but I think the answer here is we need to write some tests for this stuff :) I don't see why we shouldn't support prerelease scala versions.

File: lib/src/main/scala/org/typelevel/scalacoptions/ScalaVersion.scala [hunk 3]
```diff
@@ -44,6 +44,17 @@ object ScalaVersion {
   val V3_3_0   = ScalaVersion(3, 3, 0)
   val V3_3_1   = ScalaVersion(3, 3, 1)
 
+  private val versionRegex = raw"""(\d+)\.(\d+)\.(\d+)""".r
```
Comment (TonioGela) [inline lib/src/main/scala/org/typelevel/scalacoptions/ScalaVersion.scala:47]: [Yak Shaving Alert] Can I suggest something like this? I think we will do another PR before either Scala 4 will come out or any minor/hotfix of any existing scala version will pass number 99.
```suggestion
  private val versionRegex = raw"""(2|3)\.(\d\d?)\.(\d\d?)""".r
```

**Comment by SethTisue** [comment]
Scala version numbers aren't just `a.b.c`, they're also `a.b.c-Mx` and `a.b.c-RCx` and `a.b.c-bin-SHA` and `a.b.c-pre-SHA` and `a.b.c-bin-SHA-SNAPSHOT` and `a.b.c-pre-SHA-SNAPSHOT`. Assuming only `a.b.c` is likely to cause problems in the Scala community build, make it difficult for project maintainers to test Scala nightlies and PR snapshots, and so on.

Might the existing methods in sbt's `CrossVersion` class (https://www.scala-sbt.org/1.x/api/sbt/librarymanagement/CrossVersion$.html) be usable here?

**Comment by TonioGela** [comment]
> Might the existing methods in sbt's `CrossVersion` class (https://www.scala-sbt.org/1.x/api/sbt/librarymanagement/CrossVersion$.html) be usable here?

This could be a good idea indeed.

File: lib/src/main/scala/org/typelevel/scalacoptions/ScalaVersion.scala (hunk 2 above)
Comment (joan38) [reply lib/src/main/scala/org/typelevel/scalacoptions/ScalaVersion.scala]: I added some tests and support for `-RC` or `-M..` versions. Let me know what you think @armanbilge

**Comment by TonioGela** [comment]
@joan38 have you tried the Seth's suggestion of leveraging sbt's `CrossVersion`?

File: lib/src/test/scala/org/typelevel/scalacoptions/ScalaVersionSuite.scala [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+  property("fromString parse a version") {
+    forAll(versionGen, versionGen, versionGen) {
+      (currentMaj: Long, currentMin: Long, currentPatch: Long) =>
+        val version = ScalaVersion.fromString(s"$currentMaj.$currentMin.$currentPatch")
+        assertEquals(version, Right(ScalaVersion(currentMaj, currentMin, currentPatch)))
+    }
+  }
+
+  property("fromString parse an RC version") {
```
Comment (armanbilge) [inline lib/src/test/scala/org/typelevel/scalacoptions/ScalaVersionSuite.scala:33]: Can we also add tests for versions like:
- `3.4.0-RC1-bin-20231106-f61026d-NIGHTLY`
- `2.13.12-bin-80514f7`

**Comment by joan38** [comment]
> @joan38 have you tried the Seth's suggestion of leveraging sbt's `CrossVersion`?

Sorry I totally not see the comments.
Do we want to have a dependency on SBT?
This project is used in other build tools like mill.

File: lib/src/test/scala/org/typelevel/scalacoptions/ScalaVersionSuite.scala (hunk 4 above)
Comment (joan38) [reply lib/src/test/scala/org/typelevel/scalacoptions/ScalaVersionSuite.scala:33]: Added

**Review by armanbilge — APPROVED** [review]

**Comment by TonioGela** [comment]
> > @joan38 have you tried the Seth's suggestion of leveraging sbt's `CrossVersion`?
> 
> Sorry I totally not see the comments. Do we want to have a dependency on SBT? This project is used in other build tools like mill.

It makes sense indeed

**Comment by joan38** [comment]
Thanks guys for all the great suggestions! Excited to use this.
